File: src/main/java/com/dxp/sip/bus/fun/DispatchHandlerContext.java
package com.dxp.sip.bus.fun;


import com.dxp.sip.codec.sip.SipMethod;
import com.dxp.sip.util.ClassScanner;

import java.io.IOException;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * gb-sip 请求分发
 *
 * @author carzy
 * @date 2020/8/14
 */
public class DispatchHandlerContext {

    private static final Map<SipMethod, HandlerController> CONTROLLER_MAP = new ConcurrentHashMap<>(256);
    private static String ALLOW_METHOD = "";

    private DispatchHandlerContext() {
    }

    static {
        try {
            init();
        } catch (IOException | ClassNotFoundException | InstantiationException | IllegalAccessException e) {
            e.printStackTrace();
        }
    }

    public static HandlerController method(SipMethod method) {
        return CONTROLLER_MAP.get(method);
    }

    public static String allowMethod() {
        return ALLOW_METHOD;
    }

    public static void init() throws IOException, ClassNotFoundException, InstantiationException, IllegalAccessException {
        final Set<Class<?>> classes = ClassScanner.doScanAllClasses("com.dxp.sip.bus.fun.controller");
        for (Class<?> aClass : classes) {
            if (HandlerController.class.isAssignableFrom(aClass)) {
                addHandlerController((HandlerController) newClass(aClass));
            }
        }

        ALLOW_METHOD = CONTROLLER_MAP.keySet().stream().map(SipMethod::name).collect(Collectors.joining(","));
    }

    private static void addHandlerController(HandlerController o) {
        if (CONTROLLER_MAP.containsKey(o.method())) {
            throw new IllegalArgumentException("handlerController has be created.");
        } else {
            CONTROLLER_MAP.put(o.method(), o);
        }
    }

    private static <T> T newClass(Class<T> tClass) throws IllegalAccessException, InstantiationException {
        return tClass.newInstance();
    }

}

File: src/main/java/com/dxp/sip/bus/fun/DispatchHandlerContext.kt
package com.dxp.sip.bus.`fun`

import com.dxp.sip.codec.sip.SipMethod
import com.dxp.sip.util.ClassScanner.doScanAllClasses
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors

/**
 * gb-sip 请求分发
 *
 * @author carzy
 * @date 2020/8/14
 */
object DispatchHandlerContext {

    private val CONTROLLER_MAP: MutableMap<SipMethod?, HandlerController> = ConcurrentHashMap(256)
    private var ALLOW_METHOD = ""

    fun method(method: SipMethod): HandlerController? {
        return CONTROLLER_MAP[method]
    }

    fun allowMethod(): String {
        return ALLOW_METHOD
    }

    @Throws(IOException::class, ClassNotFoundException::class, InstantiationException::class, IllegalAccessException::class)
    fun init() {
        val classes = doScanAllClasses("com.dxp.sip.bus.fun.controller")
        for (aClass in classes) {
            if (HandlerController::class.java.isAssignableFrom(aClass)) {
                addHandlerController(newClass(aClass) as HandlerController)
            }
        }
        ALLOW_METHOD = CONTROLLER_MAP.keys.stream().map { obj: SipMethod? -> obj!!.name() }.collect(Collectors.joining(","))
    }

    private fun addHandlerController(o: HandlerController) {
        require(!CONTROLLER_MAP.containsKey(o.method())) { "handlerController has be created." }
        CONTROLLER_MAP[o.method()] = o
    }

    @Throws(IllegalAccessException::class, InstantiationException::class)
    private fun <T> newClass(tClass: Class<T>): T {
        return tClass.newInstance()
    }

    init {
        try {
            init()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }
}